toolchain/sem_ir/typed_insts.h

```diff
@@ -612,16 +655,19 @@ struct ValueOfInitializer {
 struct VarStorage {
   static constexpr auto Kind = InstKind::VarStorage.Define("var");
 
+  // TODO: Make this more specific.
   Parse::NodeId parse_node;
   TypeId type_id;
   NameId name_id;
 };
 
-// HasParseNodeMember<T> is true if T has a `Parse::NodeId parse_node` field.
-template <typename T, typename ParseNodeType = Parse::NodeId T::*>
+// HasParseNodeMember<T> is true if T has a `U parse_node` field,
+// where `U` extends `Parse::NodeId`.
+template <typename T, typename NodeIdType = Parse::NodeId>
 inline constexpr bool HasParseNodeMember = false;
 template <typename T>
-inline constexpr bool HasParseNodeMember<T, decltype(&T::parse_node)> = true;
+inline constexpr bool HasParseNodeMember<
+    T, decltype(Parse::NodeId(std::declval<T>().parse_node))> = true;
```

Comment:
Maybe it's better to check for inheritance rather than explicit convertibility? Not a strong preference though.
```suggestion
    T, std::is_base_of_v<Parse::NodeId, decltype(T::parse_node)>> = true;
```

Reply:
Accepted and fixed.